### PR TITLE
Implement Linux Transport

### PR DESCRIPTION
**Pull request by tmds**
This implements a Linux Transport for Kestrel based on the available information the Kestrel repo and the corefxlab pipeline implementation.

@davidfowl @halter73 could you do a review? The TransferThread.ReadFromSocket and TransferThread.WriteToSocket is where the socket and the pipelines meet. I know you are quite busy, so if I can get your input on the pipeline (ab)usage that would be great. If you can spare some more time, feel free to comment on the entire PR.

### CODE REVIEW

File: src/Tmds.Kestrel.Linux/TransportThread.TSocket.cs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            TypeAccept      = 0x00,
+            TypeClient      = 0x40,
+            TypePipe        = 0x80,
+
+            TypeMask        = 0xC0,
+
+            Stopping         = 0x100
+        }
+
+        class TSocket : IReadableAwaiter, IWritableAwaiter, IConnectionInformation
```
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportThread.TSocket.cs]: I don't think we expect anyone to implement these interfaces directly, but cool 😄

File: src/Tmds.Kestrel.Linux/TransportThread.cs [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+                tsocket.PipeReader = connectionContext.Output;
+
+                _sockets.TryAdd(key, tsocket);
+
+                ReadFromSocket(tsocket, connectionContext.Input);
+                WriteToSocket(tsocket, connectionContext.Output);
+            }
+        }
+
+        private async void WriteToSocket(TSocket tsocket, IPipeReader reader)
```
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: I would simplify this method to instead call Advance once.

File: src/Tmds.Kestrel.Linux/TransportThread.cs [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+            finally
+            {
+                CleanupSocket(tsocket, SocketShutdown.Receive);
+            }
+        }
+
+        private unsafe PosixResult TryReceive(Socket socket, int availableBytes, ref WritableBuffer buffer)
+        {
+            void* pointer;
+            if (!buffer.Memory.TryGetPointer(out pointer))
```
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: For sync reads and writes, just pin the span.

```C#
var span = buffer.Memory.Span;
fixed (byte* pointer = &span.DangerousGetPinnableReference())
{
    IOVector ioVector = new IOVector { Base = pointer, Count = new UIntPtr((uint)span.Length) };
    return socket.TryReceive(&ioVector, 1);
}
```

File: src/Tmds.Kestrel.Linux/TransportOptions.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+using System;
+
+namespace Tmds.Kestrel.Linux
+{
+    public class TransportOptions
+    {
+        public int ThreadCount { get; set; } = ProcessorThreadCount;
+
+        private static int ProcessorThreadCount
```
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportOptions.cs]: Maybe 2 * number of processors

Comment (tmds) [inline src/Tmds.Kestrel.Linux/TransportOptions.cs]: The Kestrel implementation is halving, why do you suggest doubling?

File: src/Tmds.Kestrel.Linux/TransportThread.cs (hunk 3 above)
Comment (tmds) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: I don't own a c# 7 compiler yet, but when I do I'll update it.

File: src/Tmds.Kestrel.Linux/TransportThread.cs [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+
+            var pipeReadKey = _pipeEnds.ReadEnd.DangerousGetHandle().ToInt32();
+            TSocket pipeReadSocket;
+            _sockets.TryRemove(pipeReadKey, out pipeReadSocket);
+            _pipeEnds.Dispose();
+
+            foreach (var kv in _sockets)
+            {
+                var tsocket = kv.Value;
+                tsocket.PipeReader.CancelPendingRead();
```
Comment (tmds) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: @davidfowl I am not sure why I am calling it but it seemed like the appropriate thing to do looking at the code in corefxlab. Wouldn't the read return for sure anyhow?
Perhaps it has to do with ensuring all client sockets are closed when StopAsync finishes? I am not enforcing this atm.

File: src/Tmds.Kestrel.Linux/TransportThread.cs [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+                            result.ThrowOnError();
+                        }
+                    }
+                    catch
+                    {
+                        buffer.Commit();
+                        throw;
+                    }
+                }
+                writer.Complete();
```
Comment (tmds) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: @davidfowl I was wondering whether I should pass an exception in more cases to Complete. Also, perhaps some of these cases (Transport Stop in particular) warrant a "connection reset by peer" close. I will keep an eye on how the corefxlab/kestrel implementations progress and adjust accordingly.

File: src/Tmds.Kestrel.Linux/TransportThread.cs (hunk 5 above)
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: You need to call it to wake up the ReadAsync call in the case of a pending read.

File: src/Tmds.Kestrel.Linux/TransportOptions.cs (hunk 4 above)
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportOptions.cs]: Bad reasons to do with libuv and measuring. Netty uses 2x number of logical cores by default.

File: src/Tmds.Kestrel.Linux/TransportThread.cs (hunk 3 above)
Comment (davidfowl) [inline src/Tmds.Kestrel.Linux/TransportThread.cs]: Get VS2017 and update yourself 😄

**Comment by davidfowl** [comment]
This is pretty darn cool. Do you have any feedback for us? I want to get to a place where we can start looking at the performance of kestrel on native transports like this one.

**Comment by tmds** [comment]
> Do you have any feedback for us?

It's great Kestrel will support pluggable Transports. The abstractions for pipelines and transport are :+1: 
Two things I missed when writing this Transport (I mentioned these elsewhere already):
- support non-contiguous write on the WritableBuffer to reduce the number recvmsg kernel calls
- a Transport being responsible to handle multiple listens to reduce the number of threads

> I want to get to a place where we can start looking at the performance of kestrel on native transports like this one.

I am looking forward to this as well. I have no idea what the perf difference is atm.
When the Transport abstraction is added to Kestrel and benchmarks are setup. If it is doable, please include this one too. You can pick it of the MyGet feed.

**Comment by tmds** [comment]
Forgot this one:
- Transport providing sendfile support

**Pull request merged**